packages/vite/src/node/fsUtils.ts

```diff
@@ -0,0 +1,416 @@
+import fs from 'node:fs'
+import path from 'node:path'
+import type { ResolvedConfig } from './config'
+import {
+  isInNodeModules,
+  normalizePath,
+  safeRealpathSync,
+  tryStatSync,
+} from './utils'
+
+export interface FsUtils {
+  existsSync: (path: string) => boolean
+  isDirectory: (path: string) => boolean
+
+  tryResolveRealFile: (
+    path: string,
+    preserveSymlinks?: boolean,
+  ) => string | undefined
+  tryResolveRealFileWithExtensions: (
+    path: string,
+    extensions: string[],
+    preserveSymlinks?: boolean,
+  ) => string | undefined
+  tryResolveRealFileOrType: (
+    path: string,
+    preserveSymlinks?: boolean,
+  ) => { path?: string; type: 'directory' | 'file' } | undefined
+
+  onFileAdd?: (file: string) => void
+  onFileUnlink?: (file: string) => void
+  onDirectoryAdd?: (file: string) => void
+  onDirectoryUnlink?: (file: string) => void
+}
+
+// An implementation of fsUtils without caching
+export const commonFsUtils: FsUtils = {
+  existsSync: fs.existsSync,
+  isDirectory,
+
+  tryResolveRealFile,
+  tryResolveRealFileWithExtensions,
+  tryResolveRealFileOrType,
+}
+
+const cachedFsUtilsMap = new WeakMap<ResolvedConfig, FsUtils>()
+export function getFsUtils(config: ResolvedConfig): FsUtils {
+  let fsUtils = cachedFsUtilsMap.get(config)
+  if (!fsUtils) {
+    if (config.command !== 'serve' || !config.server.fs.cachedChecks) {
+      // cached fsUtils is only used in the dev server for now, and only when the watcher isn't configured
+      // we can support custom ignored patterns later
+      fsUtils = commonFsUtils
+    } /* TODO: Enabling for testing, we need to review if this guard is needed
```

Comment:
I'd be fine uncommenting this to add a guard, but if we do stablize the feature, I think we should figure out getting `ignored` to work.

If we do support `ignored`, the flow could maybe be more direct, in that we can cache the fs function calls directly?

Reply:
I think we may not need this guard at the end. If the user is ignoring a certain folder or using `watch: null`, I think we could say that he needs to restart the server when adding or removing files.

Reply:
Hmm thinking about it, I don't think we invalidate the module graph for files that aren't watched, so a reload wouldn't be useful either. So maybe your idea is fine 🤔